### ISSUE TITLE
--storable-state not working in version 2.0.3

### ISSUE THREAD

**Issue by rexroni** [skvadrik/re2c#306]
My scanner started segfaulting when my system's `re2c` upgraded to 2.0.3.  Some investigation led me to see that:
- `YYSETSTATE` is never being called in the generated code
- the resulting switch statement where `YYGETSTATE` is called looks really wrong:

Input code:
```c
    // if we are continuing, restore jump to most recent re2c state
    if(scanner->continuing){
        cursor = scanner->cursor;
        marker = scanner->marker;
        yyaccept = scanner->accept;
        yych = scanner->yych;
        // clear the "continuing" flag
        scanner->continuing = 0;
        // jump into the correct state
        /*!getstate:re2c*/
    }
```

build commands:

```
re2c -W -Werror --storable-state -o scan.gen.c scan.c
gcc -g -O0 scan.gen.c -Wall -Wextra -Werror -o scan
```

generated code (with re2c 1.3, built from source):
```c
    // if we are continuing, restore jump to most recent re2c state
    if(scanner->continuing){
        cursor = scanner->cursor;
        marker = scanner->marker;
        yyaccept = scanner->accept;
        yych = scanner->yych;
        // clear the "continuing" flag
        scanner->continuing = 0;
        // jump into the correct state
        switch (YYGETSTATE()) {
default: goto yy0;
case 0: goto yyFillLabel0;
case 1: goto yyFillLabel1;
}

    }
```

generated code (with re2c 2.0.3, built from source):
```c
    // if we are continuing, restore jump to most recent re2c state
    if(scanner->continuing){
        cursor = scanner->cursor;
        marker = scanner->marker;
        yyaccept = scanner->accept;
        yych = scanner->yych;
        // clear the "continuing" flag
        scanner->continuing = 0;
        // jump into the correct state
        switch (YYGETSTATE()) {
default:
	goto yy0;
}

    }
```

----------------------------------------------------------------
Full text of scan.c:
```c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

typedef struct {
    char *buffer;
    size_t len;
    // continued scan start position
    const char* cursor;
    // continued scan backup position
    const char* marker;
    // clean scan start position
    const char* start;
    // previous scanner start position (start of the last token)
    const char* old_start;
    // part of continued scan state
    int state;
    bool accept;
    char yych;
    // for automatic continue handling
    bool continuing;
} scanner_t;

void scanner_init(scanner_t *scanner){
    // start position at beginning of buffer
    scanner->start = scanner->buffer;

    // nothing to continue
    scanner->continuing = false;
}

typedef enum {
    MODE_AAA,
    MODE_BBB,
} mode_t;

typedef enum {
    TOKEN_INVALID,
    TOKEN_AAA,
    TOKEN_BBB,
} token_t;

void scan(scanner_t *scanner, mode_t mode,
                bool *more, char **token, size_t *len, int *type){
#   define YYGETSTATE()  scanner->state
#   define YYSETSTATE(s) { \
        scanner->state = s; \
        scanner->cursor = cursor; \
        scanner->marker = marker; \
        scanner->accept = yyaccept; \
        scanner->yych = yych; \
    }
#   define YYSKIP() ++cursor
    // check before dereference.  Not efficient, but simple and always correct.
#   define YYPEEK() 0; \
                if(cursor == limit){ \
                    scanner->continuing = 1; \
                    *more = true; \
                    return; \
                }else \
                    yych = *(cursor)
#   define YYBACKUP() marker = cursor;
#   define YYRESTORE() cursor = marker;

    // instead of throwing an error, pass an error to the parser
#   define INVALID_TOKEN_ERROR { *type = TOKEN_INVALID; goto done; }

    *more = false;

    const char* cursor;
    const char* marker = NULL;
    const char* limit = scanner->buffer + scanner->len;
    unsigned yyaccept;
    (void)yyaccept;
    char yych = 0;

    // if we are continuing, restore jump to most recent re2c state
    if(scanner->continuing){
        cursor = scanner->cursor;
        marker = scanner->marker;
        yyaccept = scanner->accept;
        yych = scanner->yych;
        // clear the "continuing" flag
        scanner->continuing = 0;
        // jump into the correct state
        /*!getstate:re2c*/
    }
    // otherwise start in whatever scanner mode the parser has set for us
    else{
        yyaccept = 0;
        cursor = scanner->start;
        switch(mode){
            case MODE_AAA:  goto mode_aaa;
            case MODE_BBB:  goto mode_bbb;
        }
    }

    /*!re2c
        re2c:yyfill:enable = 0;
        re2c:flags:input = custom;
        re2c:define:YYCTYPE = char;

        eol             = "\r"?"\n";
        sp              = " ";
        literal         = "{"[0-9]+"}\r\n";
        tag             = [^(){%*+"\x00-\x1F\x7F\\ ]{1,32};
        atom_spec       = [(){%*"\]\\ ];
        atom            = [^(){%*"\x00-\x1F\x7F\]\\ ]{1,32};
        astr_atom_spec  = [(){%*"\\ ];
        astr_atom       = [^(){%*"\x00-\x1F\x7F\]\]\\ ]{1,32};
        wild_atom_spec  = [(){"\\ ];
        wild_atom       = [^(){"\x00-\x1F\x7F\\ ]{1,32};
        flag            = "\\"[^(){%*"\x00-\x1F\x7F\]\\ ]+;
        num             = [0-9]+;
        qstring         = ( [^"\x00\r\n\\] | ("\\"[\\"]) ){1,32};
        text_spec       = [ [\]];
        text_atom       = [^\x00\r\n [\]]+;
        qchar           = [^"\x00\r\n\\] | "\\\\" | "\\\"";
        nz_num          = [1-9][0-9]*;
    */

mode_aaa:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        'aaa'           { *type = TOKEN_AAA; goto done; }
    */

mode_bbb:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        'bbb'           { *type = TOKEN_BBB; goto done; }
    */

    size_t start_offset, end_offset;
done:
    // mark everything done until here
    scanner->old_start = scanner->start;
    scanner->start = cursor;

    // get the token bounds
    // this is safe; start and old_start are always within the buffer
    start_offset = (size_t)(scanner->old_start - scanner->buffer);
    end_offset = (size_t)(scanner->start - scanner->buffer);
    *token = scanner->buffer + start_offset;
    *len = end_offset - start_offset;
}

int main(int argc, char **argv){
    (void)argc;
    (void)argv;
    int type;
    char *token;
    size_t len;
    bool more;

    char *buffer = "AAABBB;";

    scanner_t scanner = {
        .buffer = buffer,
        .start = buffer,
        .len = 2,
    };

    // scan for AAA (less than a whole token is available)
    scan(&scanner, MODE_AAA, &more, &token, &len, &type);
    if(!more) return 1;

    // scan for AAA (a whole token is available)
    scanner.len = 4;
    scan(&scanner, MODE_AAA, &more, &token, &len, &type);
    if(more) return 2;
    if(type != TOKEN_AAA) return 3;
    if(len != 3) return 4;
    if(strncmp("AAA", token, 3)) return 5;

    {
        // fork the scanner state
        scanner_t scanner_copy = scanner;
        // confirm that the scanner mode is working
        scan(&scanner_copy, MODE_AAA, &more, &token, &len, &type);
        if(more) return 6;
        if(type != TOKEN_INVALID) return 7;
    }

    // scan for BBB (less than a whole token is available)
    scan(&scanner, MODE_BBB, &more, &token, &len, &type);
    if(!more) return 8;

    // scan for BBB (a whole token is available)
    scanner.len = 7;
    scan(&scanner, MODE_BBB, &more, &token, &len, &type);
    if(more) return 9;
    if(type != TOKEN_BBB) return 10;
    if(len != 3) return 11;
    if(strncmp("BBB", token, 3)) return 12;

    scan(&scanner, MODE_BBB, &more, &token, &len, &type);
    if(more) return 13;
    if(type != TOKEN_INVALID) return 14;

    return 0;
}
```

**Comment by skvadrik** [comment]
Hi @rexroni, this is indeed a regression. The unexpected thing about your example is that it uses `--storable-state` in combination with `re2c:yyfill:enable = 0;`. This was not tested before, and the `--storable-state` option implicitly assumed the case when `YYFILL` interrupts the lexer loop and returns to the caller. It just happened to work until the recent cleanup in 2.0 broke it. I will try to get it back working and certainly add a test.

**Comment by skvadrik** [comment]
If you temporarily comment out `re2c:yyfill:enable = 0;` and compare the code generated with re2c before and after 2.0, you'll see that the code is the same modulo formatting issues (of course, it won't compile, but it shows the point).

**Comment by rexroni** [comment]
Great!  Thank you.  In the mean time I will probably just fix up my code to work with this implicit assumption.

**Comment by rexroni** [comment]
Look like I can restore the previous behavior (which relies entirely on the check in `YYPEEK()`) using:
```c
#define YYFILL(len)      (void)(len)
#define YYLESSTHAN(len)  false
```

**Comment by skvadrik** [comment]
> Look like I can restore the previous behavior

Right, this workaround restores the previous behaviour.

I'm trying to understand your example better, and it seems to me that you need to handle the case when the input appears in chunks (not all available at once), and you are implementing manually much of the functionality of `YYFILL` by overriding `YYPEEK` to check for the end of input and, if necessary, save the lexer state and return to the caller. I think this doesn't work (and didn't work with old re2c), because you don't have a label where to return: if you look at the generated code,`yyFillLabel` occurrences do not match `YYPEEK` occurrences, which means that the lexer will return to the wrong point when it continues scanning.

Cases like this are meant to be handled using `YYFILL`, because then re2c inserts labels that are used to resume lexer from where it left off. See also the storable state [example](https://re2c.org/manual/manual_c.html#storable-state). Note that `YYFILL` can be used either with [bounds-checking](https://re2c.org/manual/manual_c.html#bounds-checking-with-padding) or [EOF rule](https://re2c.org/manual/manual_c.html#eof-rule) methods, the latter is a bit simpler. I'd be happy to help with rewriting the code to use `YYFILL`, and I think it will be both more efficient and more correct than overriding `YYPEEK`, but I don't fully understand your use case and how exactly you receive the input. Are there some specific restrictions or reasons to not use `YYFILL`?

**Comment by rexroni** [comment]
Hm... Well it was quite a while ago that I wrote this scanner.

My use case is an IMAP client/server, where packets of data come in from the network, so there's no guarantees that the packet framing has anything to do with token boundaries.  I suspect that I was nervous about using `YYFILL` because if you were in the wrong scan mode (IMAP has an awfully complex grammar), and you might get a "need more bytes" from the scanner if an IMAP command from the client didn't terminate quite right, which you would have been able to detect had you just looked at the next character of the command.  You can actually see this behavior in the example I wrote, where if you implement YYFILL/YYLESSTHAN correctly, with the input text of `AAABBB;` you get a `TOKEN_AAA`, a `TOKEN_BBB`, and then a `more=true` scan.  The `YYPEEK` strategy I used will return with ` TOKEN_INVALID` on the third scan instead.

But the truth is, I didn't really know how to do error handling in parsers when I wrote that, and the only sane way to do error handling in an IMAP parser is to wait to return errors until you see a line break, so the distinction above is really pretty pointless.

I appreciate your offer to help me rewrite it to use YYFILL!  Let me take a shot at it first (probably monday or tuesday) and you can tell me if I got it right?

**Comment by skvadrik** [comment]
@rexroni Sure, ping me when you are ready. The [storable state example](https://re2c.org/manual/manual_c.html#storable-state) is right about sending and receiving packets, so it should be applicable to your case.

Regarding this bug, I think it's better to disallow the use of `--storable-state` together with `re2c:yyfill:enable = 0;`, as it is almost certainly incorrect. This is backwards incompatible, but perhaps it is better to break in such cases.

**Comment by rexroni** [comment]
@skvadrik thank you for your help here.  After trying to port this back to my main codebase I hit a bug that I could not resolve with the `YYFILL` strategy, and that is that `YYLESSTHAN`/`YYFILL` are being called with values that are incorrectly long for what I need.  For example, if I have one pattern in a state that either matches a token of repeated B's (`[Bb]{1,32}`) or an end-of-line marker (`"\r"?"\n"`), I see a `YYLESSTHAN(32)` that has to be satisfied before the lexer will even look at the contents of the buffer.  In practice, that means that my lexer might refuse to tell me that an IMAP client has finished sending a command, while the IMAP client is blocking waiting for me to respond to the command it sent me.  I'll include the full text of this example below (it fails the same under re2c v1.3 as under v2.0.3).

As a side note, I think my `YYPEEK` strategy actually *is* technically correct.  If you look at the generated code, I store the `cursor`, `marker`, `accept`, and `yych` values with every `YYSETSTATE` and I restore them to their previous state before I call `/*!getstate:re2c*/`.  I don't see how that could fail, nor I have I ever seen any odd behavior in the year that I've been playing with it.  I was digging through the docs to switch to the `YYFILL` strategy and I ran across this line, which I think is where I originally got the idea for my `YYPEEK` strategy:

> Generic API was added in version 0.14. It is intentionally designed to give the user as much freedom as possible in redefining the input model and the semantics of different actions performed by the generated code. As an example, one can override YYPEEK to check for the end of input before reading the input character, or do some logging, etc.
---------------------------------
Problem input:
```c
mode_bbb:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        "\r"?"\n"       { *type = TOKEN_EOL; goto done; }
        [Bb]{1,32}      { *type = TOKEN_BBB; goto done; }
    */
```
Resulting output:
```c
mode_bbb:

    
#line 168 "scan.gen.c"

	YYSETSTATE(1);
	if (YYLESSTHAN(32)) YYFILL(32);  // will never tokenize the string "BBB\n"
```
---------------------------------
Full text of scan.c
```c// example illustrating problem with YYFILL, returns 5 instead of 0

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <stdint.h>

typedef struct {
    char *buffer;
    size_t len;
    // continued scan start position
    const char* cursor;
    // continued scan backup position
    const char* marker;
    // clean scan start position
    const char* start;
    // previous scanner start position (start of the last token)
    const char* old_start;
    // part of continued scan state
    int state;
    bool accept;
    char yych;
    // for automatic continue handling
    bool continuing;
} scanner_t;

void scanner_init(scanner_t *scanner){
    // start position at beginning of buffer
    scanner->start = scanner->buffer;

    // nothing to continue
    scanner->continuing = false;
}

typedef enum {
    MODE_AAA,
    MODE_BBB,
} mode_t;

typedef enum {
    TOKEN_INVALID,
    TOKEN_AAA,
    TOKEN_BBB,
    TOKEN_EOL,
} token_t;

void scan(scanner_t *scanner, mode_t mode,
                bool *more, char **token, size_t *len, int *type){
#   define YYGETSTATE()  scanner->state
#   define YYSETSTATE(s) { \
        scanner->state = s; \
        scanner->cursor = cursor; \
        scanner->marker = marker; \
        scanner->accept = yyaccept; \
        scanner->yych = yych; \
    }
#   define YYSKIP() ++cursor
#   define YYPEEK() *(cursor)
#   define YYBACKUP() marker = cursor;
#   define YYRESTORE() cursor = marker;
#   define YYFILL(len) \
        do { \
            *more = true; \
            scanner->continuing = 1; \
            return; \
        } while(0)
#   define YYLESSTHAN(len) ((uintptr_t)(limit - cursor) < (len))

    // instead of throwing an error, pass an error to the parser
#   define INVALID_TOKEN_ERROR { *type = TOKEN_INVALID; goto done; }

    *more = false;

    const char* cursor;
    const char* marker = NULL;
    const char* limit = scanner->buffer + scanner->len;
    unsigned yyaccept;
    (void)yyaccept;
    char yych = 0;

    // if we are continuing, restore jump to most recent re2c state
    if(scanner->continuing){
        cursor = scanner->cursor;
        marker = scanner->marker;
        yyaccept = scanner->accept;
        yych = scanner->yych;
        // clear the "continuing" flag
        scanner->continuing = 0;
        // jump into the correct state
        /*!getstate:re2c*/
    }
    // otherwise start in whatever scanner mode the parser has set for us
    else{
        yyaccept = 0;
        cursor = scanner->start;
        switch(mode){
            case MODE_AAA:  goto mode_aaa;
            case MODE_BBB:  goto mode_bbb;
        }
    }

    /*!re2c
        re2c:flags:input = custom;
        re2c:define:YYCTYPE = char;
    */

mode_aaa:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        'aaa'           { *type = TOKEN_AAA; goto done; }
    */

mode_bbb:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        "\r"?"\n"       { *type = TOKEN_EOL; goto done; }
        [Bb]{1,32}      { *type = TOKEN_BBB; goto done; }
    */

    size_t start_offset, end_offset;
done:
    // mark everything done until here
    scanner->old_start = scanner->start;
    scanner->start = cursor;

    // get the token bounds
    // this is safe; start and old_start are always within the buffer
    start_offset = (size_t)(scanner->old_start - scanner->buffer);
    end_offset = (size_t)(scanner->start - scanner->buffer);
    *token = scanner->buffer + start_offset;
    *len = end_offset - start_offset;
}

int main(int argc, char **argv){
    (void)argc;
    (void)argv;
    int type;
    char *token;
    size_t len;
    bool more;

    char *buffer = "AAABBB\n";

    scanner_t scanner = {
        .buffer = buffer,
        .start = buffer,
        .len = 7,
    };

    // scan for AAA
    scan(&scanner, MODE_AAA, &more, &token, &len, &type);
    if(more) return 1;
    if(type != TOKEN_AAA) return 2;
    if(len != 3) return 3;
    if(strncmp("AAA", token, 3)) return 4;

    // scan for BBB
    // (this returns 5 due to the YYLESSTHAN(32) call in the generated code)
    scan(&scanner, MODE_BBB, &more, &token, &len, &type);
    if(more) return 5;
    if(type != TOKEN_AAA) return 6;
    if(len != 3) return 7;
    if(strncmp("BBB", token, 3)) return 8;

    // scan for EOL
    scan(&scanner, MODE_BBB, &more, &token, &len, &type);
    if(more) return 9;
    if(type != TOKEN_EOL) return 10;
    if(len != 1) return 11;
    if(strncmp("\n", token, 1)) return 12;

    return 0;
}
```

**Comment by skvadrik** [comment]
> YYLESSTHAN/YYFILL are being called with values that are incorrectly long for what I need.

This is a known limitation of the default end-of-input checking mechanism, you should use [EOF rule](https://re2c.org/manual/manual_c.html#eof-rule), as the [storable state example](https://re2c.org/manual/manual_c.html#storable-state) does. Here is a patch that fixes your example (note that I had to fix `TOKEN_AAA` -> `TOKEN_BBB`, I think it was a copy-paste error), and below is the explanation:

```diff
--- a.re   2020-09-15 07:27:54.741736005 +0100
+++ b.re   2020-09-15 07:27:38.533735353 +0100
@@ -40,6 +40,7 @@
     TOKEN_AAA,
     TOKEN_BBB,
     TOKEN_EOL,
+    TOKEN_EOF,
 } token_t;
.
 void scan(scanner_t *scanner, mode_t mode,
@@ -100,12 +101,14 @@
     /*!re2c
         re2c:flags:input = custom;
         re2c:define:YYCTYPE = char;
+        re2c:eof = 0;
     */
.
 mode_aaa:
.
     /*!re2c
         *               { INVALID_TOKEN_ERROR; }
+        $               { *type = TOKEN_EOF; goto done; }
         'aaa'           { *type = TOKEN_AAA; goto done; }
     */
.
@@ -113,6 +116,7 @@
.
     /*!re2c
         *               { INVALID_TOKEN_ERROR; }
+        $               { *type = TOKEN_EOF; goto done; }
         "\r"?"\n"       { *type = TOKEN_EOL; goto done; }
         [Bb]{1,32}      { *type = TOKEN_BBB; goto done; }
     */
@@ -158,7 +162,7 @@
     // (this returns 5 due to the YYLESSTHAN(32) call in the generated code)
     scan(&scanner, MODE_BBB, &more, &token, &len, &type);
     if(more) return 5;
-    if(type != TOKEN_AAA) return 6;
+    if(type != TOKEN_BBB) return 6;
     if(len != 3) return 7;
     if(strncmp("BBB", token, 3)) return 8;
```

The default EOF handling mechanism in re2c is [bounds-checking with padding](https://re2c.org/manual/manual_c.html#bounds-checking-with-padding). In order to make the lexer run faster, it has to minimize the number of end-of-input checks. So re2c does the following trick: it computes strongly connected components in the lexer (these are groups of DFA states with loops where you can get from any state to any other state --- roughly speaking, loops), and places a check at the entry of each strongly connected component. And instead of checking for one character, it checks for multiple characters at a time --- this allows the lexer skip subsequent checks and run uninterrupted for a while. To make this work, it is necessary to provide some padding at the end of input; the length of padding is `YYMAXFILL`, and it should not form a valid lexeme (zeroes usually are a good choice).

This does make the lexer run faster, but it is problematic or doesn't work in cases like yours, where input comes in chunks and the lexer is interrupted to wait for more input. Therefore a different EOF handling mechanism is needed, namely EOF rule `$`. It requires exactly one sentinel character at the end of buffer (zero in case of `re2c:eof = 0`). It uses a different trick to minimize the number of end-of-input checks: the lexer only checks for EOF when it reads the sentinel character (or one of the characters that fall into the same conditional branch).

> As a side note, I think my YYPEEK strategy actually is technically correct. If you look at the generated code, I store the cursor, marker, accept, and yych values with every YYSETSTATE and I restore them to their previous state before I call /*!getstate:re2c*/. I don't see how that could fail, nor I have I ever seen any odd behavior in the year that I've been playing with it.

You do set pointers, state and other variables, that's right. But the state that you save does not correspond to the DFA state where the lexer is interrupted, it corresponds to some other state that has `YYFILL` (and if you don't have `YYFILL`, there is no such state). There is no label corresponding to the state where the lexer was interrupted. In our example, consider what happens if the lexer is interrupted after reading two `A`s. Where will it resume?

> I was digging through the docs to switch to the YYFILL strategy and I ran across this line, which I think is where I originally got the idea for my YYPEEK strategy:

Right, overriding `YYPEEK` allows to have custom EOF handling methods, but it doesn't work in this case because the lexer state that you backup and restore with `YYSETSTATE` does not correspond 1:1 to the DFA states (unless you use EOF rule, in which case you don't need the `YYPEEK` magic). If you compare the code generated for the original example with an old re2c version, and the code generated with the above patch, you will see that the latter contains a label for _each_ state, because the interrupt may happen in each state.

**Comment by rexroni** [comment]
> But the state that you save does not correspond to the DFA state where the lexer is interrupted

Ok, I looked back at the generated code, and I see this now.  Thank you.

> In order to make the lexer run faster, it has to minimize the number of end-of-input checks.

So originally I was thinking that this was going to put a burden on my code to support the mechanics of a sentinel and padded buffers for a performance gain that I don't even care about... but then I realized that all of saveable state is also a performance gain I don't care about it.  I'm just going to drop that flag, and lots of things will get simpler for me : )

And FWIW, not supporting storable state with yyfill:enable=0 makes more sense to me now.

Thank you for your help!

**Comment by rexroni** [comment]
> how will you handle the case when a packet arrives piece by piece?

Just restart from the beginning of a token.  That's very little work lost, and it's very rare that it will happen in practice.  In practice, the large majority of IMAP commands will be entire lines that fit into a single network buffer.

> EOF rule should be much simpler than padding

The EOF rule *is* a form of padding.  Well... looking at a different parser I have laying around I once wrote `YYPEEK` to return a `\0` conditionally when the end of the buffer was reached, which doesn't require padding.  But more importantly, relying on a sentinel would require me to sanitize user inputs, and it would become a pretty obnoxious error recovery scenario since all other forms of invalid input are handled very neatly by the lexer and parser.

**Comment by skvadrik** [comment]
Ah, I see. Just one more note:

> relying on a sentinel would require me to sanitize user inputs

I'm not sure if you this is what you mean, but EOF rule does not require the absence of sentinel in the middle of the string (it is different from the [simple sentinel method](http://re2c.org/manual/manual_c.html#using-sentinel-symbol)). So, for example, if you have `re2c:eof = 0;`, it is totally fine to have input string containing `\0` in the middle. The lexer will stop at `\0` to check if it is the end of input of just an ordinary `\0`. So with EOF rule you can use the lexer on any inputs, as long as you append `re2c:eof` symbol at the end.

**Comment by rexroni** [comment]
Oh, that's neat.  Does it behave semantically different though?  Like if I try to lex user input with a \0 in it:

```
AAA\0BBB\0
```

It would spit out a `TOKEN_AAA`, I'd trim the input and lex on

```
\0BBB\0
```

Then what would it do?

**Comment by skvadrik** [comment]
Assuming that the lexer is in mode `mode_bbb` with the rules shown below, the cursor points at `\0BBB\0`, and the limit points at the second `\0`, then the lexer will match default rule `*` that will consume `\0` and return `INVALID_TOKEN_ERROR`.
```c
mode_bbb:

    /*!re2c
        *               { INVALID_TOKEN_ERROR; }
        "\r"?"\n"       { *type = TOKEN_EOL; goto done; }
        [Bb]{1,32}      { *type = TOKEN_BBB; goto done; }
    */
```

The decision if `\0` is the end of input or just a regular symbol is based solely on the position of cursor relative to limit: if the encountered `\0` is at limit, then it is the end of input. Otherwise the lexer calls `YYFILL`, which can supply more input and adjust pointers. When the lexer resumes, it will re-match the character at cursor (supplied instead of `\0`), so that the net effect will be as if it matched continuous input without `\0`.

In other words, any `\0` that is before limit will be matched as a regular symbol.

**Comment by rexroni** [comment]
Ok, that all makes a lot more sense.  I have read through the EOF documentation a few times, and I think if you broke it down to this level in there it would help a lot.

Thanks again for all your help here!

**Comment by skvadrik** [comment]
@rexroni Agreed, documentation can be better, I can add the above explanation there. :)

Commit https://github.com/skvadrik/re2c/commit/520e5cd433d033125df6b72faebc4c9962315752 disallows using `-f, --storable-state` option if `YYFILL is disabled`.

**Comment by skvadrik** [comment]
@rexroni FYI I updated [the docs](https://re2c.org/manual/manual_c.html#handling-the-end-of-input), hopefully they are more helpful now.

**Comment by rexroni** [comment]
@skvadrik I had a chance to look back over them... they are very clear now!  Thanks!
